programs/server/Server.cpp

```diff
@@ -838,6 +843,25 @@ int Server::main(const std::vector<std::string> & /*args*/)
 
             global_context->updateStorageConfiguration(*config);
             global_context->updateInterserverCredentials(*config);
+            if (clusters_config_has_changed)
+            {
+                LOG_DEBUG(log, "Reload the connection pool of distributed tables.");
+                for (const auto & db : DatabaseCatalog::instance().getDatabases())
+                {
+                    if (!db.second->canContainDistributedTables())
+                        continue;
+                    for (auto iterator = db.second->getTablesIterator(context()); iterator->isValid(); iterator->next())
+                    {
+                        StoragePtr table = iterator->table();
+                        if (!table)
+                            continue;
+                        auto distributed_table = dynamic_cast<StorageDistributed *>(table.get());
+                        if (!distributed_table)
+                            continue;
+                        distributed_table->updateConnectionPool();
```

Comment:
Actually this is a bit uncontrolled, when I though about this my concern was that cluster definition can be changed at any time, and once it was changed it should atomically sync all existing pending blocks for distributed send and then re-create all directory monitors.

Not sure, but maybe making this asynchronously is enough...